File: Course_1-Algorithmic_Toolbox/Week-5/Excercise_Challenges/2_primitive_calculator/primitive_calculator.py
def optimal_sequence_dp(n):
    cal = [0] * (n + 1)

    for i in range(1, n+1):
        cal[i] = cal[i - 1] + 1
        if i % 2 == 0:
            cal[i] = min(cal[i], cal[i // 2] + 1)
        if i % 3 == 0:
            cal[i] = min(cal[i], cal[i // 3] + 1)

    sequence = [n]
    while n != 1:
        if n % 3 == 0 and cal[n] - 1 == cal[n // 3]:
            sequence.append(n // 3)
            n = n//3
        elif n % 2 == 0 and cal[n] - 1 == cal[n // 2]:
            sequence.append(n // 2)
            n = n//2
        else:
            sequence.append(n - 1)
            n -= 1

    return sequence[::-1]

def optimal_sequence(n):
    sequence = []
    while n >= 1:
        sequence.append(n)
        if n % 3 == 0:
            n = n // 3
        elif n % 2 == 0:
            n = n // 2
        else:
            n = n - 1
    return reversed(sequence)

if __name__ == '__main__':
    n = int(input())
    
    sequence = list(optimal_sequence_dp(n))
    print(len(sequence) - 1)
    for x in sequence:
        print(x, end=' ')

File: Course_1-Algorithmic_Toolbox/Week-5/Excercise_Challenges/2_primitive_calculator/primitive_calculator.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

vector<int> optimal_sequence_dp(int n)
{
    vector<int> sequence;
    int cal[n + 1];
    cal[0] = 0;

    for(int i = 1; i <= n; ++i){
        cal[i] = cal[i - 1] + 1;
        if(i % 2 == 0){
            cal[i] = min(cal[i/2] + 1, cal[i]);
        }
        if(i % 3 == 0){
            cal[i] = min(cal[i/3] + 1, cal[i]);
        }
    }

    sequence.push_back(n);
    while(n != 1){
        if(n % 3 == 0 && cal[n] - 1 == cal[n / 3]){
            sequence.push_back(n / 3);
            n /= 3;
        }
        else if(n % 2 == 0 && cal[n] - 1 == cal[n / 2]){
            sequence.push_back(n / 2);
            n /= 2;
        }
        else{
            sequence.push_back(n - 1);
            n--;
        }
    }

    reverse(sequence.begin(), sequence.end());
    return sequence;
}

vector<int> optimal_sequence(int n)
{
    vector<int> sequence;
    while(n >= 1){
        sequence.push_back(n);
        if(n % 3 == 0){
            n /= 3;
        }
        else if(n % 2 == 0){
            n /= 2;
        }
        else{
            n -= 1;
        }
    }

    reverse(sequence.begin(), sequence.end());
    return sequence;
}

int main(void)
{
    int n;
    cin >> n;

    vector<int> sequence = optimal_sequence_dp(n);
    cout << sequence.size() - 1 << '\n';
    for(size_t i = 0; i < sequence.size(); ++i){
        cout << sequence[i] << " ";
    }
    cout << endl;

    return 0;
}
